README.md
# LearnTV Release v0.0.20

## Installation & Upgrade 

To install or upgrade LearnTV utility scripts v0.0.20, you should run the [install](https://raw.githubusercontent.com/raviu/learntv-releases/v0.0.20/install.sh) script. To do that, you may either download and run the script manually, or use the following cURL or Wget command:

```
curl -o- https://raw.githubusercontent.com/raviu/learntv-releases/v0.0.20/install.sh | bash
```

```
wget -qO- https://raw.githubusercontent.com/raviu/learntv-releases/v0.0.20/install.sh | bash
```


## Version Archive 

For the latest version always go here: https://github.com/raviu/learntv-releases
### v0.0.19
https://github.com/raviu/learntv-releases/tree/v0.0.19
### v0.0.18
https://github.com/raviu/learntv-releases/tree/v0.0.18
### v0.0.17
https://github.com/raviu/learntv-releases/tree/v0.0.17
### v0.0.16
https://github.com/raviu/learntv-releases/tree/v0.0.16
### v0.0.15
https://github.com/raviu/learntv-releases/tree/v0.0.15
### v0.0.14
https://github.com/raviu/learntv-releases/tree/v0.0.14
### v0.0.13
https://github.com/raviu/learntv-releases/tree/v0.0.13
### v0.0.12
https://github.com/raviu/learntv-releases/tree/v0.0.12
### v0.0.11
https://github.com/raviu/learntv-releases/tree/v0.0.11
### v0.0.10
https://github.com/raviu/learntv-releases/tree/v0.0.10
### v0.0.9
https://github.com/raviu/learntv-releases/tree/v0.0.9
### v0.0.8
https://github.com/raviu/learntv-releases/tree/v0.0.8
### v0.0.7
https://github.com/raviu/learntv-releases/tree/v0.0.7
### v0.0.6
https://github.com/raviu/learntv-releases/tree/v0.0.6
### v0.0.5
https://github.com/raviu/learntv-releases/tree/v0.0.5
### v0.0.4
https://github.com/raviu/learntv-releases/tree/v0.0.4
### v0.0.3
https://github.com/raviu/learntv-releases/tree/v0.0.3
### v0.0.2
https://github.com/raviu/learntv-releases/tree/v0.0.2
### v0.0.1
https://github.com/raviu/learntv-releases/tree/v0.0.1


CHANGELOG.md

 **2022-06-27**: release.sh add changelog to release repo


 **2022-06-27**: thumbnail-extract.sh: add find command to run script for multiple files into help


 **2022-06-27**: thumbnail-extract.sh: add -qmin 1 so that -q:v 1 actually works


 **2022-06-27**: thumbnail-extract.sh: add image output format flag


 **2022-06-21**: release.sh: add thumnail-extract directory to release


 **2022-06-21**: release.sh: add thumnail-extract directory to release


 **2022-06-21**: thumbnail-extract: add thumbnail-extract to install script and release script


 **2022-06-21**: thumbnail-extract.sh: adding thumbnail extraction script


 **2021-09-23**: channel-starter: playout logs should be per channel identifer + key, not just channel identifier


 **2021-09-23**: channel-starter: make log message clearer


 **2021-09-23**: channel-starter: move pids and logs to directories


 **2021-09-22**: channel-starter: only act on pids file if it exists


 **2021-09-22**: channel-starter: only remove pids file if it exists


 **2021-09-22**: channel-starter: only use port to identify ffmpeg processes


 **2021-09-22**: iptv-live: save ffmpeg output in /var/log/$TODAY-iptv-live-ffmpeg-$URL.log


 **2021-09-22**: release.sh: cron-template is now channel-starter


 **2021-09-22**: install: channel-starter needs to be installed in /usr/local/bin


 **2021-09-22**: install: fix some issues with channel-starter.sh


 **2021-09-22**: iptv-live.sh: add check to terminnate if M3U file is missing!"


 **2021-09-22**: channel-starter: add kill flag functionality and log iptv-live.sh output to log file in BASE_LOCATION


 **2021-09-22**: install: remove cron-template and add channel-starter


 **2021-09-22**: iptv-live: add version in comment


 **2021-09-22**: youtube-live: add version in comments


 **2021-09-22**: channel-starter: new channel starter script to start channels


 **2021-09-22**: cront-template: add configurable ports for cron template


 **2021-09-22**: release: add cront-template.sh to release


 **2021-09-22**: cron: rename to cron-template.sh and new cron-template that kills existing processes bbefore starting playback


 **2021-09-22**: install.sh: add cron_template.sh


 **2021-09-22**: iptv-live: remove new line in first line of script


 **2021-09-22**: release: add version archive to README


 **2021-09-22**: release: detect existing tags and report latest and all tags


 **2021-09-22**: release: test branch name


 **2021-09-22**: release: remove branch on push


 **2021-09-22**: release: use $REPO


 **2021-09-22**: install: add installer script and release automation for release repo


 **2021-09-22**: playout: update scripts with placeholders for version and updated date


 **2021-09-22**: README: add installation and upgrade instructions


 **2021-09-21**: youtube-live.sh: whitespace fix


 **2021-09-21**: youtube-live.sh: YouTube live streaming requires -re for real time streaming. Removed -qscale as it is redundant. Added flvflags no_duration_filesize. Removed threads as ffmpeg can auto detect this value to use


 **2021-09-21**: cron: first line must be shebang


 **2021-09-20**: cron.sh: update contributors


 **2021-09-20**: README: update output descriptions


 **2021-09-20**: playout: update playlist finished output


 **2021-09-20**: playout: add playlist finished output


 **2021-09-20**: make-m3u: validate month input is in supported format


 **2021-09-02**: cron: add sample cron file


 **2021-09-20**: playout: make sure current time updates at start of loop


 **2021-09-20**: READEM: update how to read script output


 **2021-09-20**: README: update how to read output instructions


 **2021-09-20**: README: update how to read script output


 **2021-09-20**: playout: update README with insturctions how to read output


 **2021-09-20**: Fix README


 **2021-09-20**: fix README


 **2021-09-20**: iptv-live.sh: added IPTV playout script. Cleaned up echo output


 **2021-09-07**: Merge pull request #2 from raviu/raviu-patch-1


 **2021-09-07**: Update youtube-live.sh


 **2021-09-07**: Update youtube-live.sh


 **2021-09-07**: Update youtube-live.sh


 **2021-09-07**: Update youtube-live.sh


 **2021-09-06**: Merge pull request #1 from raviu/enable_streaming


 **2021-09-06**: youtube-live: use gnu date and also added -s or start flag to specify the playback start time for the script


 **2021-09-05**: Update youtube-live.sh


 **2021-09-05**: Update README.md


 **2021-09-05**: Merge branch 'main' into enable_streaming


 **2021-09-05**: youtube-live: enable ffmpeg streaming


 **2021-09-06**: README: update usage to show correct m3u filename format


 **2021-09-05**: README: update readme for youtube-live script


 **2021-09-05**: playout/youtube-live: add initial youtube live script with skipping


 **2021-09-05**: make-m3u: add delimiter to EXTINF that is more robust with more entropy


 **2021-09-05**: youtube-live: uncomment echo of ffmpet command


 **2021-09-05**: youtube-live: set playlist start time to 04:59:59 to give cron a chance to start exactly at 05:00:00


 **2021-09-05**: youtube-live.sh: add a guard to not allow streaming before scheduled start time


 **2021-09-05**: make-m3u: add -m/--month argument that is mandatory and make -v video path that is optional with a default


 **2021-09-04**: make-m3u: fix bug where DAY is not incremented when file already exists


 **2021-09-04**: README/tests: update tests to use the media argument and update readme


 **2021-09-04**: make-m3u: pass media directory location as an argument


 **2021-09-04**: .DS_Store banished!


 **2021-09-04**: README: rename old readme so it does not show up as github readme


 **2021-09-04**: README: rename old readme so it does not show up as github readme


 **2021-09-04**: README: add usage output in readme


 **2021-09-04**: README: get rid of link to output folder as it will not exist on git


 **2021-09-04**: tests: run all tests inside output folder for easy cleanup



install.sh
#!/usr/bin/env bash

{ # force full scrip to download first

echo "                                                                  ..:-==+*##+. .          .-++=-.     "
echo "                                                      .=***###%%%%@@@@@@@%%%@@@@@#=      +@@@@@@@#:   "
echo "                                                     :%@@%%%%###***+++++===+#**+*@@-    =@@+==++%@%   "
echo "                                                     *@%=======+++++++++++++++++=%@#   .@@#=+++*@@-   "
echo "                                                     *@%=++==++++++++****###%++++*@@:  #@%+++++@@+    "
echo "                                                     -@@#*##%%%#++++*@@@%%%@@+++++@@* =@@+++++%@#     "
echo "        .::::::::::::::::::::::::::::::::::::::::::::.=#%%%##@@%=+++*@@.   %@#=+++#@@:@@#=+++*@@:     "
echo "       .::#%#%+:::*%#%%##*::=%#%%#%-::+%#%%###+:::#%##:.=*** #@%++++#@%    =@@+++++@@@@@+++++@@+      "
echo "        ::%@@@*:::#@@@@@@@::*@@@@@@+::*@@@@@@@@#::@@@@+:#@@@.*@%++++#@%     @@#=++=#@@@#=++=#@%       "
echo "        ::%@@@*:-:#@@@@%%#::#@@@@@@*::*@@@@*@@@@=:%@@@%:#@@@.*@%++++#@%     +@@+++++@@@+++++@@=       "
echo "        ::%@@@*:-:#@@@%:::::%@@@@@@#::*@@@@:#@@@+:%@@@@-*@@@.#@%++++#@%     .@@*+++=%@#++++#@@        "
echo "        ::%@@@*:-:#@@@%::::-@@@%%@@@-:*@@@@:#@@@+:%@@@@**@@@.#@#=+++#@%      *@%++++*@+++++@@+        "
echo "        ::%@@@*:-:#@@@@**+:=@@@#*@@@-:*@@@@*@@@@=:%@@@@@#@@@.#@#=+++#@%      :@@*++++*++++*@@:        "
echo "        ::%@@@*:-:#@@@@@@#:+@@@+*@@@=:*@@@@@@@%=::%@@@@@@@@@.#@%++++#@@       #@%+++++++++%@%         "
echo "        ::%@@@*:-:#@@@@@@*:*@@@=+@@@*:*@@@@#@@@#-:%@@@@@@@@@.#@%++++*@@.      :@@*++++++++@@+         "
echo "        ::%@@@*:-:#@@@%---:#@@@=+@@@#:*@@@@:%@@@+:%@@%%@@@@@.#@%++++*@@.       #@%++++++++@@-         "
echo "        ::%@@@*:::#@@@%::::%@@@@@@@@%:*@@@@:#@@@+:%@@%+@@@@@.#@%++++#@@.       -@@+++++++*@@.         "
echo "        ::%@@@*--:#@@@%----@@@@@@@@@@:*@@@@:#@@@+:%@@%-@@@@@.#@%++++*@@.        %@#=+++++#@%          "
echo "        ::%@@@@@@=*@@@@@@@+@@@@+=@@@@-*@@@@:#@@@+:%@@%:#@@@@.#@%++++*@@.        =@@++++++%@#          "
echo "        ::%@@@@@@=#@@@@@@@*@@@@-:@@@@+*@@@@:%@@@+:@@@@:=@@@@.#@#=+++*@@.        .@@#=+++=%@*          "
echo "       .::*######-+#######*####-:*###++####:*###=:*##*::####.#@%++++#@@.         +@@*++++@@+          "
echo "        ::::::::::::::::::::::::::::::::::::::::::::::::::::::#@@@@@@%=           =#@@@@@%*.          "
echo "                                                               .:--:.                .::.             "
echo "                                                                                                      "
echo "                                                                                                      "
echo "                                                                                                      "
echo "                                                                                                      "
echo "                                                                                                      "

BASE_URL="https://raw.githubusercontent.com/raviu/learntv-releases/v0.0.20"
VERSION="v0.0.20"

if command -v curl &> /dev/null 
then
    DOWNLOADER=curl
else 
    if command -v wget &> /dev/null
    then
        DOWNLOADER=wget
    fi
fi

if [ -z "$DOWNLOADER" ]; then
    echo "To continue install either wget or curl is required. Please install one of these utilities and try again."
    echo ""
    exit 1
else 
    echo "Downloading scripts to /usr/local/bin"
    if [ "$DOWNLOADER" == "curl" ]; then 
        curl -s $BASE_URL/make-m3u/make-m3u.sh -o /usr/local/bin/make-m3u.sh 
        curl -s $BASE_URL/thumbnail-extract/thumbnail-extract.sh -o /usr/local/bin/thumbnail-extract.sh
        curl -s $BASE_URL/playout/youtube-live.sh -o /usr/local/bin/youtube-live.sh 
        curl -s $BASE_URL/playout/iptv-live.sh -o /usr/local/bin/iptv-live.sh 
        curl -s $BASE_URL/playout/channel-starter.sh -o /usr/local/bin/channel-starter.sh 
    else 
        wget -q $BASE_URL/make-m3u/make-m3u.sh -O /usr/local/bin/make-m3u.sh
        wget -q $BASE_URL/thumbnail-extract/thumbnail-extract.sh -O /usr/local/bin/thumbnail-extract.sh
        wget -q $BASE_URL/playout/youtube-live.sh -O /usr/local/bin/youtube-live.sh 
        wget -q $BASE_URL/playout/iptv-live.sh -O /usr/local/bin/iptv-live.sh
        wget -q $BASE_URL/playout/channel-starter.sh -O /usr/local/bin/channel-starter.sh 
    fi 
    chmod +x /usr/local/bin/make-m3u.sh
    chmod +x /usr/local/bin/thumbnail-extract.sh
    chmod +x /usr/local/bin/youtube-live.sh
    chmod +x /usr/local/bin/iptv-live.sh
    chmod +x /usr/local/bin/channel-starter.sh
fi

CUR_DIR=$(pwd)
echo "Installed make-m3u.sh to /usr/local/bin"
echo "Installed thumbnail-extract.sh to /usr/local/bin"
echo "Installed youtube-live.sh to /usr/local/bin"
echo "Installed iptv-live.sh to /usr/local/bin"
echo "Installed channel-starter.sh to /usr/local/bin"
echo "Installation of LearnTV utilities $VERSION complete"

} # force full scrip to download first


make-m3u/make-m3u.sh
#!/usr/bin/env bash 

# Contributors: http://github.com/raviu/
# Last Updated: 27-06-2022

# Transform long options to short ones
for arg in "$@"; do
  shift
  case "$arg" in
    "--help")       set -- "$@" "-h" ;;
    "--csv")        set -- "$@" "-c" ;;
    "--month")      set -- "$@" "-m" ;;
    "--videopath")  set -- "$@" "-v" ;;
    *)              set -- "$@" "$arg"
  esac
done

# Display help 
function display_help() {
    cat << EOF
make-m3u v0.0.20 generates m3u playlist given CSV file as input

Usage (items inside [] are optional arguments):

./make-m3u.sh -c|--csv <file> -m|--month <month> [-v|--videopath <media directory path>]

Example

./make-m3u.sh --csv /home/learntv/2021-06-meditation-v2.csv --month JUN

Options:
-----------------------------
-c or --csv         CSV file path
-m or --month       Month (must be one of: jan, feb, mar, apr, jun, jul, aug, sep, oct, nov, dec)
-v or --videopath   Media directory location to search in. Deafults to /media/data/video/
-h or --help        Display help
EOF
}

# Global Variables 
MEDIA_PATH=/media/data/video/
NOT_FOUND_FILE=NOT_FOUND.txt
DUPLICATES_FILE=DUPLICATES_FOUND.txt

# Command flag and argument processing
while getopts  ":c:m:v:h" opt; do
    case $opt in 
        c)
            CSV=$OPTARG 
            ;;
        m)
            MONTH=$OPTARG 
            ;;
        v)
            echo "Media path overridden to $OPTARG"
            MEDIA_PATH=$OPTARG 
            ;;
        h) 
            display_help
            exit 0 
            ;;
        :)
            echo "Option -$OPTARG requires an argument." >&2
            exit 1
            ;;
        \?)
            echo "Invalid option: -$OPTARG" >&2
            exit 1
            ;;
    esac
done 

# Validate Required Arguments 
shift "$(( OPTIND - 1 ))"
if [ -z "$CSV" ] || [ -z "$MONTH" ]; then
    echo "The -c and -m arguments are required. Please run ./make-m3u.sh --help for help." >&2
    exit 1
fi

months=(jan feb mar apr may jun jul aug sep oct nov dec)

if [[ ! " ${months[*]} " =~ " ${MONTH,,} " ]]; then
    echo "The -m or --month argument must be one of these: jan, feb, mar, apr, may, jun, jul, aug, sep, oct, nov, dec. Please run ./make-m3u.sh --help for help." >&2
    exit 1
fi

# Cleanup  
echo "Cleaning up previous $NOT_FOUND_FILE file"
rm $NOT_FOUND_FILE 2>/dev/null
echo "Cleaning up previous $DUPLICATES_FILE file"
rm $DUPLICATES_FILE 2>/dev/null

# Program
DAY=0
echo "Making m3u files from $CSV"
while IFS=, read -r broadcast_time duration video_file_name; do

    if [[ "$duration" == "#day" ]]; then
        ((DAY++))
        if [ -f "${MONTH,,}-$DAY".m3u ]; then
            continue
        fi
        echo "#EXTM3U" > "${MONTH,,}-$DAY".m3u
        continue
    fi

    readarray -td '' video_files < <(find -L "${MEDIA_PATH}" -type f -name "$video_file_name" -print0)
    if [ ${#video_files[@]} -eq 0 ]; then
        echo "Day $DAY "$video_file_name" cannot be found" >> "$NOT_FOUND_FILE"
        continue
    elif [ ${#video_files[@]} -gt 1 ]; then
        echo "Duplicates found for Day $DAY "$video_file_name": " >> "$DUPLICATES_FILE"
        echo "${video_files[@]}" >> "$DUPLICATES_FILE"
    fi

    video_file="${video_files[0]}"
    duration_seconds=$(echo $duration | awk -F':' '{print $1 * 60 * 60 + $2 * 60 + $3}')

    echo "#EXTINF:$duration_seconds,START_TIME:$broadcast_time" >> "${MONTH,,}-$DAY".m3u
    echo "$video_file" >> "${MONTH,,}-$DAY".m3u

done < $CSV



playout/youtube-live.sh
#!/usr/bin/env bash 

# Contributors: http://github.com/raviu/
# Last Updated: 27-06-2022
# Version: v0.0.20

# Global Variables with defaults 
VBR="2500k"                                   
FPS="30"                                       
QUAL="medium"                                  
YOUTUBE_URL="rtmp://a.rtmp.youtube.com/live2"  
START_TIME="04:59:59"

DATE_CMD="date"
if [[ $OSTYPE == 'darwin'* ]]; then
  DATE_CMD="gdate"
fi

# Transform long options to short ones
for arg in "$@"; do
  shift
  case "$arg" in
    "--help")           set -- "$@" "-h" ;;
    "--m3u")            set -- "$@" "-m" ;;
    "--quality")        set -- "$@" "-q" ;;
    "--bitrate")        set -- "$@" "-b" ;;
    "--fps")            set -- "$@" "-f" ;;
    "--key")            set -- "$@" "-k" ;;
    "--start")          set -- "$@" "-s" ;;
    *)                  set -- "$@" "$arg"
  esac
done

# Display help 
function display_help() {
    cat << EOF
youtube-live.sh v0.0.20 streams m3u playlist file to YouTube

Usage (items inside [] are optional arguments):

./youtube-live.sh.sh -m|--m3u <file> -k|--key <YouTube Key> [-q|--quality <quality string> -b|--bitrate <variable bit rate> -f|--fps <frames per second> -s|--start <start time in HH:mm:ss>]

Example

./youtube-live.sh --m3u /home/learntv/dharmavahini/2021/09/sep-4.m3u -k YOUTUBE_KEY

Options:
-----------------------------
-m or --m3u         M3U playlist file path. This argument is required. 
-k or --key         YouTube streaming key. This argument is required. 
-q or --quality     Quality string like medium. This is optional and will default to medium. 
-b or --bitrate     Variable bitrate. This is optional and will default to 2500k.
-f or --fps         Frames per second. This is optional and will default to 30.
-s or --start       Specify the start time of the playlist in 24 hour format HH:mm:ss. Defaults to 05:00:00 (5AM)
-h or --help        Display help
EOF
}

# Command flag and argument processing
while getopts  ":m:q:b:f:k:s:h" opt; do
    case $opt in 
        m)
            M3U=$OPTARG 
            PLAYLIST_DAY=`basename $OPTARG | cut -d'-' -f2 | cut -d'.' -f1`
            ;;
        q)
            QUAL=$OPTARG 
            ;;
        b)
            VBR=$OPTARG 
            ;;
        f)
            FPS=$OPTARG 
            ;;
        k)
            KEY=$OPTARG 
            ;;
        s)
            START_TIME=$OPTARG
            ;;
        h) 
            display_help
            exit 0 
            ;;
        :)
            echo "Option -$OPTARG requires an argument." >&2
            exit 1
            ;;
        \?)
            echo "Invalid option: -$OPTARG" >&2
            exit 1
            ;;
    esac
done 

# Validate Required Arguments 
shift "$(( OPTIND - 1 ))"
if [ -z "$KEY" ] || [ -z "$M3U" ]; then
    echo "The -k and -m arguments are both required. Please run ./youtube-live.sh --help for help." >&2
    exit 1
fi                                  

PLAYLIST_START_TIME=$($DATE_CMD --date "`$DATE_CMD +'%Y'`-`$DATE_CMD +'%m'`-$PLAYLIST_DAY $START_TIME" "+%s")
echo "PLAYLIST_START_TIME `$DATE_CMD -d @$PLAYLIST_START_TIME`"

PLAYCURSOR=$PLAYLIST_START_TIME
CURRENT_TIME=`$DATE_CMD +%s`
echo "CURRENT_TIME `$DATE_CMD -d @$CURRENT_TIME`"

if [[ $CURRENT_TIME -lt $((PLAYCURSOR)) ]]; then 
    echo "The playlist $M3U cannot be played before `$DATE_CMD -d @$PLAYCURSOR`! Are you trying to stream it before scheduled start time?"
    exit 1
fi

KEEP_SKIPPING=1
SKIP_NEXT=0
while read -r line; do

    CURRENT_TIME=`$DATE_CMD +%s`

    if [[ "$line" == "#EXTM3U" ]]; then
        continue
    elif [[ "$line" == \#* ]]; then 
        video_start_time=`echo $line | awk -v delimeter="START_TIME:" '{split($0,a,delimeter)} END{print a[2]}'`
        duration_seconds=`echo $line | awk -v delimeter="START_TIME:" '{split($0,a,delimeter)} END{print a[1]}' | cut -d ':' -f2 | cut -d',' -f1`
        echo "Item Starts $video_start_time"
        echo "Item Ends `$DATE_CMD -d @$((PLAYCURSOR+duration_seconds))`"
    else 
        if [[ "$SKIP_NEXT" == "1" ]]; then 
            SKIP_NEXT=0
            echo "Skipping $line"
            echo "=++++++========+++++======++++++========+++++======++++++========+++++======++++++========+++++====="  
            continue
        else 
            echo "Playing $line with skip seconds $skip_seconds"

            ffmpeg \
            -nostdin -y -re -i "$line" \
            -ss $skip_seconds \
            -vcodec libx264 -pix_fmt yuv420p -preset $QUAL -r $FPS -g $(($FPS * 2)) -b:v $VBR \
            -acodec libmp3lame -ar 44100 -b:a 712000 -bufsize 512k \
            -f flv -flvflags no_duration_filesize "$YOUTUBE_URL/$KEY"

            PLAYCURSOR=$((PLAYCURSOR+ (duration_seconds-skip_seconds)))
            skip_seconds=0
            KEEP_SKIPPING=0
            echo "Playback complete. PLAYCURSOR AT `$DATE_CMD -d @$PLAYCURSOR`"
            echo "=++++++========+++++======++++++========+++++======++++++========+++++======++++++========+++++=====" 
            continue
        fi
    fi

    if [ "$KEEP_SKIPPING" == "1" ]; then 
        if [ $CURRENT_TIME -lt $((PLAYCURSOR+duration_seconds)) ]; then 
            #echo "current time < end time"
            skip_seconds=$(( duration_seconds - ((PLAYCURSOR+duration_seconds) - CURRENT_TIME)))
            skip_seconds=$((skip_seconds<0 ? 0 : skip_seconds))
            SKIP_NEXT=0
            echo "Marking item for playback after $skip_seconds seconds. PLAYCURSOR AT `$DATE_CMD -d @$PLAYCURSOR`"
            continue
        else 
            #echo "current time > end time"
            PLAYCURSOR=$((PLAYCURSOR+duration_seconds))
            SKIP_NEXT=1
            echo "Marking item for skip. PLAYCURSOR NOW AT `$DATE_CMD -d @$PLAYCURSOR`"
            continue
        fi    
    fi

done < $M3U

echo "Playlist finished"


playout/cron-template.sh
#!/usr/bin/env bash 

# Contributors: http://github.com/raviu/, https://github.com/mettavihari
# Last Updated: 22-09-2021
# Version: v0.0.5

CHANNEL_IDENTIFIER=channel_1
PORT_360=3011
PORT_480=3012
PORT_720=3013

IPTV_LIVE_PIDS_LOCATION=/opt/dls-2/iptv-live-$CHANNEL_IDENTIFIER.pids

## Kill currently running iptv-live.sh processes based on pids file stored in IPTV_LIVE_PIDS_LOCATION
while read -r pid; do
  echo "Killing previous process $pid"
  kill -9 $pid
done < $IPTV_LIVE_PIDS_LOCATION

rm $IPTV_LIVE_PIDS_LOCATION

## Kill ffmpeg processes that may still be running 
mapfile -t FFMPEG_PORT_360 < <( pgrep -d " " -f ^ffmpeg.*$CHANNEL_IDENTIFIER.*$PORT_360.* )
mapfile -t FFMPEG_PORT_480 < <( pgrep -d " " -f ^ffmpeg.*$CHANNEL_IDENTIFIER.*$PORT_480.* )
mapfile -t FFMPEG_PORT_720 < <( pgrep -d " " -f ^ffmpeg.*$CHANNEL_IDENTIFIER.*$PORT_720.* )

FFMPEG_CHANNEL_PIDS=("${FFMPEG_PORT_360[@]}" "${FFMPEG_PORT_480[@]}" "${FFMPEG_PORT_720[@]}") 
for ffmpeg_pid in "${FFMPEG_CHANNEL_PIDS[@]}"
do
    echo "Killing ffmpeg process $ffmpeg_pid"
    kill -9 $ffmpeg_pid
done

DATE_CMD="date"
if [[ $OSTYPE == 'darwin'* ]]; then
  DATE_CMD="gdate"
fi

MONTH=`$DATE_CMD +%b`
DAY=`$DATE_CMD +'%-d'`

PIDS=()

/opt/dls-2/iptv-live.sh --m3u /home/edu/playlists/$CHANNEL_IDENTIFIER\_360/${MONTH,,}-$DAY.m3u --url 127.0.0.1:$PORT_360 & 
PIDS+=($!)

/opt/dls-2/iptv-live.sh --m3u /home/edu/playlists/$CHANNEL_IDENTIFIER\_480/${MONTH,,}-$DAY.m3u --url 127.0.0.1:$PORT_480 &
PIDS+=($!)

/opt/dls-2/iptv-live.sh --m3u /home/edu/playlists/$CHANNEL_IDENTIFIER\_720/${MONTH,,}-$DAY.m3u --url 127.0.0.1:$PORT_720 & 
PIDS+=($!)

for pid in "${PIDS[@]}"
do
    echo "$pid" >> $IPTV_LIVE_PIDS_LOCATION
done


playout/iptv-live.sh
#!/usr/bin/env bash 

# Contributors: http://github.com/raviu/
# Last Updated: 27-06-2022
# Version: v0.0.20

# Global Variables with defaults                                
START_TIME="04:59:59"

DATE_CMD="date"
if [[ $OSTYPE == 'darwin'* ]]; then
  DATE_CMD="gdate"
fi

# Transform long options to short ones
for arg in "$@"; do
  shift
  case "$arg" in
    "--help")           set -- "$@" "-h" ;;
    "--m3u")            set -- "$@" "-m" ;;
    "--url")            set -- "$@" "-u" ;;
    "--start")          set -- "$@" "-s" ;;
    *)                  set -- "$@" "$arg"
  esac
done

# Display help 
function display_help() {
    cat << EOF
iptv-live.sh v0.0.20 streams m3u playlist file to IPTV service

Usage (items inside [] are optional arguments):

./iptv-live.sh.sh -m|--m3u <file> -u|--url <streaming url> [ -s|--start <start time in HH:mm:ss> ]

Example

./iptv-live.sh --m3u /home/learntv/dharmavahini/2021/09/sep-4.m3u --url 127.0.0.1:2001

Options:
-----------------------------
-m or --m3u         M3U playlist file path. This argument is required. 
-u or --url         The streaming URL including any ports. Ex: 127.0.0.1:2001
-s or --start       Specify the start time of the playlist in 24 hour format HH:mm:ss. Defaults to 05:00:00 (5AM)
-h or --help        Display help
EOF
}

# Command flag and argument processing
while getopts  ":m:s:u:h" opt; do
    case $opt in 
        m)
            M3U=$OPTARG 
            PLAYLIST_DAY=`basename $OPTARG | cut -d'-' -f2 | cut -d'.' -f1`
            ;;
        s)
            START_TIME=$OPTARG
            ;;
        u)
            URL=$OPTARG
            ;;
        h) 
            display_help
            exit 0 
            ;;
        :)
            echo "Option -$OPTARG requires an argument." >&2
            exit 1
            ;;
        \?)
            echo "Invalid option: -$OPTARG" >&2
            exit 1
            ;;
    esac
done 

# Validate Required Arguments 
shift "$(( OPTIND - 1 ))"
if [ -z "$M3U" ] || [ -z "$URL" ]; then
    echo "The -m and -u arguments are both required. Please run ./iptv-live.sh --help for help." >&2
    exit 1
fi 

if [ ! -f "$M3U" ]; then 
    echo "Could not find playlist file $M3U. Terminating."
    exit 1
fi 

PLAYLIST_START_TIME=$($DATE_CMD --date "`$DATE_CMD +'%Y'`-`$DATE_CMD +'%m'`-$PLAYLIST_DAY $START_TIME" "+%s")
echo "PLAYLIST_START_TIME `$DATE_CMD -d @$PLAYLIST_START_TIME`"

TODAY=`$DATE_CMD +'%Y-%m-%d'`
PLAYCURSOR=$PLAYLIST_START_TIME
CURRENT_TIME=`$DATE_CMD +%s`
echo "CURRENT_TIME `$DATE_CMD -d @$CURRENT_TIME`"

if [[ $CURRENT_TIME -lt $((PLAYCURSOR)) ]]; then 
    echo "The playlist $M3U cannot be played before `$DATE_CMD -d @$PLAYCURSOR`! Are you trying to stream it before scheduled start time?"
    exit 1
fi

KEEP_SKIPPING=1
SKIP_NEXT=0
while read -r line; do

    CURRENT_TIME=`$DATE_CMD +%s`

    if [[ "$line" == "#EXTM3U" ]]; then
        continue
    elif [[ "$line" == \#* ]]; then 
        video_start_time=`echo $line | awk -v delimeter="START_TIME:" '{split($0,a,delimeter)} END{print a[2]}'`
        duration_seconds=`echo $line | awk -v delimeter="START_TIME:" '{split($0,a,delimeter)} END{print a[1]}' | cut -d ':' -f2 | cut -d',' -f1`
        echo "Item Starts $video_start_time"
        echo "Item Ends `$DATE_CMD -d @$((PLAYCURSOR+duration_seconds))`"
    else 
        if [[ "$SKIP_NEXT" == "1" ]]; then 
            SKIP_NEXT=0
            echo "Skipping $line"
            echo "=++++++========+++++======++++++========+++++======++++++========+++++======++++++========+++++====="  
            continue
        else 
            echo "Playing $line with skip seconds $skip_seconds"

            ffmpeg \
            -nostdin -re \
            -i "$line" \
            -ss $skip_seconds \
            -codec copy \
            -f mpegts udp://$URL?pkt_size=1316 >> /var/log/$TODAY-iptv-live-ffmpeg-$URL.log 2>&1

            PLAYCURSOR=$((PLAYCURSOR+ (duration_seconds-skip_seconds)))
            skip_seconds=0
            KEEP_SKIPPING=0
            echo "Playback complete. PLAYCURSOR AT `$DATE_CMD -d @$PLAYCURSOR`"
            echo "=++++++========+++++======++++++========+++++======++++++========+++++======++++++========+++++=====" 
            continue
        fi
    fi

    if [ "$KEEP_SKIPPING" == "1" ]; then 
        if [ $CURRENT_TIME -lt $((PLAYCURSOR+duration_seconds)) ]; then 
            #echo "current time < end time"
            skip_seconds=$(( duration_seconds - ((PLAYCURSOR+duration_seconds) - CURRENT_TIME)))
            skip_seconds=$((skip_seconds<0 ? 0 : skip_seconds))
            SKIP_NEXT=0
            echo "Marking item for playback after $skip_seconds seconds. PLAYCURSOR AT `$DATE_CMD -d @$PLAYCURSOR`"
            continue
        else 
            #echo "current time > end time"
            PLAYCURSOR=$((PLAYCURSOR+duration_seconds))
            SKIP_NEXT=1
            echo "Marking item for skip. PLAYCURSOR NOW AT `$DATE_CMD -d @$PLAYCURSOR`"
            continue
        fi    
    fi

done < $M3U

echo "Playlist finished"
